### PR TITLE
Signon

### PR DESCRIPTION
**Pull request by benthorner**

### CODE REVIEW

**Comment by rossjones** [comment]
Would it be possible to rebase this as I think a lot of the deletions are already done on master, it might make the PR a bit smaller.

**Comment by benthorner** [comment]
Sure - I was going to leave Andrew alone to review it, but may as well.

File: app/controllers/home_controller.rb [hunk 1]
```diff
@@ -1,27 +0,0 @@
-class HomeController < ApplicationController
```
Comment (andrewgarner) [inline app/controllers/home_controller.rb]: We shouldn't be removing this behaviour.

File: app/mailers/application_mailer.rb [hunk 2]
```diff
@@ -1,4 +0,0 @@
-class ApplicationMailer < ActionMailer::Base
```
Comment (andrewgarner) [inline app/mailers/application_mailer.rb]: I'm not sure we should remove this here. We no longer need to send emails in relation to user accounts but removing wider support for ActionMailer is something we should consider separately in the context of the whole application, not authentication.

File: app/views/home/dashboard.html.erb [hunk 3]
```diff
@@ -1,71 +0,0 @@
-<% content_for :page_title do %>
-  What's available in Publish
```
Comment (andrewgarner) [inline app/views/home/dashboard.html.erb]: We shouldn't just remove this feature. It exists for a reason, we should still be able to provide an overview of what's available in Publish when authenticating through signon.

File: app/views/home/quality.html.erb [hunk 4]
```diff
@@ -1,34 +0,0 @@
-<% content_for :page_title do %>
-    Organisation quality -
```
Comment (andrewgarner) [inline app/views/home/quality.html.erb]: We shouldn't remove this.

File: config/initializers/active_admin.rb [hunk 5]
```diff
@@ -1,293 +0,0 @@
-ActiveAdmin.setup do |config|
```
Comment (andrewgarner) [inline config/initializers/active_admin.rb]: We should only remove instances where ActiveAdmin has been used to for authentication or the management of users. Removing ActiveAdmin completely from the codebase is something we should review in a wider context.

File: spec/controllers/datasets_controller_spec.rb [hunk 6]
```diff
@@ -1,68 +0,0 @@
-require 'rails_helper'
-
-describe DatasetsController, type: :controller do
```
Comment (andrewgarner) [inline spec/controllers/datasets_controller_spec.rb]: This behaviour being tested here is unrelated to signon. We should continue to test it.

File: spec/features/create_dataset_spec.rb [hunk 7]
```diff
@@ -7,11 +7,11 @@
 
   context "when the user goes through entire flow" do
     before(:each) do
-      sign_in_user
+      sign_in_as(user)
     end
 
     it "navigates to new dataset form" do
-      expect(page).to have_current_path("/tasks")
```
Comment (andrewgarner) [inline spec/features/create_dataset_spec.rb]: The tasks behaviour should be unaffected by using signon.

File: spec/models/link_spec.rb [hunk 8]
```diff
@@ -1,31 +1,15 @@
 require 'rails_helper'
 
-describe Link, type: :model do
```
Comment (andrewgarner) [inline spec/models/link_spec.rb]: This is unrelated.

File: spec/models/user_spec.rb [hunk 9]
```diff
@@ -1,39 +0,0 @@
-require 'rails_helper'
-
-describe User do
```
Comment (andrewgarner) [inline spec/models/user_spec.rb]: How about we use the `it_behaves_like "a gds-sso user class"` example - https://github.com/alphagov/gds-sso#testing-in-your-application?

File: spec/rails_helper.rb [hunk 10]
```diff
@@ -63,7 +63,3 @@
   # arbitrary gems may also be filtered via:
   # config.filter_gems_from_backtrace("gem name")
 end
-
-def last_updated_dataset
```
Comment (andrewgarner) [inline spec/rails_helper.rb]: It would be preferable to propose changes like this in a separate pull-request.

File: app/controllers/home_controller.rb (hunk 1 above)
Comment (benthorner) [reply app/controllers/home_controller.rb]: John was happy with this one. Should be OK.

File: app/mailers/application_mailer.rb (hunk 2 above)
Comment (benthorner) [reply app/mailers/application_mailer.rb]: I'm not saying we can never send email, just that we don't need to right now. We can always add it properly (or from the history) if and when we need to.

File: config/initializers/active_admin.rb (hunk 5 above)
Comment (benthorner) [reply config/initializers/active_admin.rb]: I'm hoping we can get to a clean slate - leaving things like this lying around when they can't be used will create bloat.

File: spec/controllers/datasets_controller_spec.rb (hunk 6 above)
Comment (benthorner) [reply spec/controllers/datasets_controller_spec.rb]: This controller test was testing authorisation logic, which I've removed as it's incompatible with the new user model (also half-baked). The tests needed to go as the logic they tested is also gone.

File: spec/features/create_dataset_spec.rb (hunk 7 above)
Comment (benthorner) [reply spec/features/create_dataset_spec.rb]: Unfortunately the home page and related pages are no longer relevant, which is consistent with other GOV.UK admin apps. Jon's happy for now that we redirect to the tasks page, which is consistent with the current behaviour when signing-in.

File: spec/rails_helper.rb (hunk 10 above)
Comment (benthorner) [reply spec/rails_helper.rb]: I agree - I intended to do it as part of my let! PR but forgot it. I will try to remember better in future.

File: app/mailers/application_mailer.rb (hunk 2 above)
Comment (rossjones) [reply app/mailers/application_mailer.rb]: Please leave it in now.

File: app/controllers/home_controller.rb (hunk 1 above)
Comment (andrewgarner) [reply app/controllers/home_controller.rb]: I've asked @nicholsj about this. The dashboard is a quick way for admins to get an overview but will need to be looked at in the future.

File: app/models/ability.rb [hunk 11]
```diff
@@ -1,11 +0,0 @@
-class Ability
-  include CanCan::Ability
-
-  def initialize(user)
-    user ||= User.new
-
-    can [:read, :update], Dataset do |dataset|
-      user.creator_of_dataset?(dataset) || user.in_organisation?(dataset.organisation)
```
Comment (andrewgarner) [inline app/models/ability.rb]: I've asked @nicholsj for a steer on this. He said that we definitely need permissions for editing (and creating) datasets and that they should be tied to the owner organisation of the dataset.

File: app/views/home/dashboard.html.erb (hunk 3 above)
Comment (nicholsj) [reply app/views/home/dashboard.html.erb]: Sorry to jump in  but hope it's useful. I'm not sure I did know this would be removing the "What's in Publish" dashboard (sorry if I gave that impression at some point). While that ashboard obvs needs some discussion/work, I think we should keep it for now. I can't see how it's related to SignOn?

File: app/views/home/quality.html.erb (hunk 4 above)
Comment (nicholsj) [reply app/views/home/quality.html.erb]: Similar to my comment on the dashboard - let's keep this. We'll have to work out access controls for some of these pages, which I assume is why it's come up now, but that can come later

File: app/views/home/index.html.erb [hunk 12]
```diff
@@ -1,35 +0,0 @@
...
-      </h1>
-      <p class="lede">
-        You must create an account to use this service.
-      </p>
-      <p class="lede">
-        If you already have an account, sign in to publish a new dataset
-        or update an existing one.
-      </p>
-
-      <a class="button dgu-masthead__button-get-started" href="<%= new_user_session_path %>" role="button">
```
Comment (andrewgarner) [inline app/views/home/index.html.erb]: I've asked @nicholsj about this one too. His steer was that the public facing page should remain and the signon part for now should be limited to changing the "Sign in" button to use the authentication flow provided by Signon instead of our own UI.
